pkg/controller/pipelinegates/reconciler.go

```diff
@@ -0,0 +1,189 @@
+package pipelinegates
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	console "github.com/pluralsh/console-client-go"
+	generated "github.com/pluralsh/deployment-operator/generated/client/clientset/versioned"
+	"github.com/pluralsh/deployment-operator/pkg/client"
+	"github.com/pluralsh/deployment-operator/pkg/controller/service"
+	"github.com/pluralsh/deployment-operator/pkg/ping"
+	"github.com/pluralsh/deployment-operator/pkg/websocket"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/pluralsh/deployment-operator/internal/utils"
+)
+
+func init() {
+	Local = false
+}
+
+var (
+	Local = false
+)
+
+type GateReconciler struct {
+	ConsoleClient   client.Client
+	Config          *rest.Config
+	Clientset       *kubernetes.Clientset
+	GenClientset    *generated.Clientset
+	GateCache       *client.Cache[console.PipelineGateFragment]
+	GateQueue       workqueue.RateLimitingInterface
+	UtilFactory     util.Factory
+	discoveryClient *discovery.DiscoveryClient
+	pinger          *ping.Pinger
+}
+
+func NewGateReconciler(consoleClient client.Client, config *rest.Config, refresh time.Duration, clusterId string) (*GateReconciler, error) {
+	service.DisableClientLimits(config)
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	gateCache := client.NewCache[console.PipelineGateFragment](refresh, func(id string) (*console.PipelineGateFragment, error) {
+		return consoleClient.GetClusterGate(id)
+	})
+
+	gateQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	f := utils.NewFactory(config)
+
+	cs, err := f.KubernetesClientSet()
+	if err != nil {
+		return nil, err
+	}
+
+	genClientset, err := generated.NewForConfig(config)
```

Comment:
You could also use runtime client like we did in restore controller (`k8sClient`). Then perhaps we would need to generate informers etc.